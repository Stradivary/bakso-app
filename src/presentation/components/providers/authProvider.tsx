import { Database } from "@/data/models/supabase.types";
import { signInUser } from "@/domain/services/authService";
import { supabase } from "@/domain/services/supabaseService";
import { notifications } from "@mantine/notifications";
import { Session } from "@supabase/supabase-js";
import {
  createContext,
  PropsWithChildren,
  useCallback,
  useContext,
  useEffect,
  useMemo,
  useState,
} from "react";

type UserDetails = Database["public"]["Tables"]["user_profiles"]["Row"];

export interface AuthContextType {
  session: Session | null;
  user: UserDetails | null;
  isLoading: boolean;
  error: string | null;
  login: (
    name: string,
    role: string,
    latitude: number,
    longitude: number,
  ) => Promise<void>;
  logout: () => Promise<void>;
  isAuthenticated: boolean;
}

export const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [session, setSession] = useState<Session | null>(null);
  const [user, setUser] = useState<UserDetails | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);
  const [initialized, setInitialized] = useState(false);

  // Fetch user profile data
  const fetchUserProfile = useCallback(async (userId: string) => {
    try {
      const { data, error: fetchError } = await supabase
        .from("user_profiles")
        .select("*")
        .eq("id", userId)
        .single();

      if (fetchError) throw fetchError;

      if (data) {
        setUser(data);
      }
    } catch (error) {
      console.error("Error fetching user profile:", error);
      setError("Failed to fetch user profile");
    }
  }, []);

  // Initialize auth state
  useEffect(() => {
    const initializeAuth = async () => {
      try {
        // Check for existing session
        const {
          data: { session: existingSession },
          error: sessionError,
        } = await supabase.auth.getSession();

        if (sessionError) {
          console.error("failed to set session, log in again");
          throw sessionError;
        }

        if (existingSession) {
          setSession(existingSession);
          await fetchUserProfile(existingSession.user.id);
        }
      } catch (error) {
        console.error("Error initializing auth:", error);
        setError("Failed to initialize authentication");
      } finally {
        setIsLoading(false);
        setInitialized(true);
      }
    };

    initializeAuth();
  }, [fetchUserProfile]);
  const login = useCallback(
    async (name: string, role: string, latitude: number, longitude: number) => {
     
      try {
        setIsLoading(true);
        const { session: newSession, error: signInError } = await signInUser(
          name,
          role,
          latitude,
          longitude,
        );
        if (signInError) throw signInError;

        setSession(newSession);
        if (newSession) {
          await fetchUserProfile(newSession.user.id);
        }
        setError(null);
      } catch (error) {
        console.error("Error during login:", error);
        setError("Failed to login");
      } finally {
        setIsLoading(false);
      }
    },
    [fetchUserProfile],
  );

  const logout = useCallback(async () => {
    try {
      const { error: signOutError } = await supabase.auth.signOut({
        scope: "local",
      });

      if (signOutError) throw signOutError;
      setSession(null);
      setUser(null);
    } catch (error) {
      console.error("Error during logout:", error);
      setError("Failed to logout");
    }
  }, []);

  // Storage change observer
  useEffect(() => {
    if (!initialized) return;

    const handleStorageChange = (event: StorageEvent) => {
      const currentSession = event.newValue;
      const previousSession = event.oldValue;

      if (event.key === "abangbakso-session") {
        if (!previousSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session injection detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
          return;
        }

        if (!currentSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session removal detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
          return;
        }
        // If session was changed externally
        if (currentSession !== previousSession) {
          notifications.show({
            title: "Warning",
            message: "Possible session tampering detected",
            color: "orange",
          });
          logout();
          setError("Session security violation detected");
        }
      }
    };

    // Add storage event listener
    window.addEventListener("storage", handleStorageChange);

    return () => {
      window.removeEventListener("storage", handleStorageChange);
    };
  }, [initialized, logout, session?.user.id]);

  useEffect(() => {
    if (!initialized) return;

    const {
      data: { subscription },
    } = supabase.auth.onAuthStateChange(async (event, newSession) => { 

      if (!newSession?.user.id) return;
      if ((newSession?.user.id && newSession?.user.id) && newSession?.user.id !== session?.user.id) {
        return;
      }
      switch (event) {
        case "INITIAL_SESSION":
          break;
        case "SIGNED_IN":
        case "TOKEN_REFRESHED":
        case "USER_UPDATED":
          setSession(newSession);
          if (newSession) {
            await fetchUserProfile(newSession.user.id);
          }
          break;
        case "SIGNED_OUT":
          setSession(null);
          setUser(null);
          break;
      }
    });

    return () => {
      subscription.unsubscribe();
    };
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [initialized, fetchUserProfile]);

  const isAuthenticated = !!session?.user?.id && !isLoading;

  const values = useMemo(
    () => ({
      session,
      user,
      isLoading,
      error,
      login,
      logout,
      isAuthenticated,
    }),
    [session, user, isLoading, error, login, logout, isAuthenticated],
  );

  return <AuthContext.Provider value={values}>{children}</AuthContext.Provider>;
};

export const AuthContext = createContext<AuthContextType | undefined>(
  undefined,
);

export const useAuth = () => {
  const context = useContext(AuthContext);
  if (context === undefined) {
    throw new Error("useAuth must be used within an AuthProvider");
  }
  return context;
};
