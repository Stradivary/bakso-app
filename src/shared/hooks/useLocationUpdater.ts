import { useEffect } from "react";
import { LatLng } from "leaflet";
import { useLocation } from "./useLocation";

export const UPDATE_INTERVAL = 2000;

export function useLocationUpdater({
  userId,
  userRole,
  updateLocation,
}: {
  userId: string;
  userRole: string;
  updateLocation: (userId: string, location: LatLng) => void;
}) {
  const { location } = useLocation();

  useEffect(() => {
    if (userRole !== "seller") {
      return;
    }

    const updateInterval = setInterval(() => {
      updateLocation(
        userId,
        new LatLng(location?.latitude ?? 0, location?.longitude ?? 0),
      );
    }, UPDATE_INTERVAL);

    return () => clearInterval(updateInterval);
  }, [
    userRole,
    updateLocation,
    location?.latitude,
    location?.longitude,
    userId,
  ]);

  return location;
}
