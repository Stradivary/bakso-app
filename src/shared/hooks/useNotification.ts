import { useCallback, useState } from "react";

export interface Notification {
  buyer_id: string | null;
  created_at: string | null;
  expiry_at: string | null;
  id: string;
  seller_id: string | null;
  is_read: boolean | null;
  buyer_name?: string;
}

export function useNotifications(
  userId: string | null | undefined,
  initialNotifications: Notification[] = [],
) {
  const [notifications, setNotifications] = useState<Notification[]>(initialNotifications); // Use initialNotifications
  const [unreadCount, setUnreadCount] = useState(
    initialNotifications.filter(note => !note.is_read).length
  );

  const markAsRead = useCallback(async (notificationId: string) => {
    if (!userId) return;

    try {
      // Optimistically update the UI
      setNotifications(prev =>
        prev.map(n =>
          n.id === notificationId ? { ...n, is_read: true } : n
        )
      );
      setUnreadCount(prev => Math.max(0, prev - 1));
    } catch (error) {
      console.error("Error in markAsRead:", error);
    }
  }, [userId]);

  return {
    notifications,
    unreadCount,
    markAsRead,
  };
}