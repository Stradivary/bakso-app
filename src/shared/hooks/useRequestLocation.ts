// useRequestLocation.ts
import { useState } from "react";

interface LocationState {
  latitude: number | null;
  longitude: number | null;
  error: string | null;
  isLoading: boolean;
}

export const useRequestLocation = () => {
  const [state, setState] = useState<LocationState>({
    latitude: null,
    longitude: null,
    error: null,
    isLoading: true,
  });

  const requestLocation = () => {
    if (!navigator.geolocation) {
      setState((prev) => ({
        ...prev,
        error: "Browser anda tidak mendukung geolocation",
        isLoading: false,
      }));
      return;
    }

    setState((prev) => ({ ...prev, isLoading: true }));

    navigator.geolocation.getCurrentPosition(
      (position) => {
        setState((prev) => ({
          ...prev,
          latitude: position.coords.latitude,
          longitude: position.coords.longitude,
          error: null,
          isLoading: false,
        }));
      },
      (error) => {
        setState({
          latitude: null,
          longitude: null,
          error: error.message,
          isLoading: false,
        });
      },
    );
  };

  return { ...state, requestLocation };
};
