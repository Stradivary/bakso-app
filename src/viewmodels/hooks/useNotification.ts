import { Notification } from "@/models/Notification.types";
import { useCallback, useState } from "react";

export function useNotifications(
  userId: string | null | undefined,
  initialNotifications: Notification[] = [],
) {
  const [notifications, setNotifications] =
    useState<Notification[]>(initialNotifications); // Use initialNotifications
  const [unreadCount, setUnreadCount] = useState(
    initialNotifications.filter((note) => !note.is_read).length,
  );

  const markAsRead = useCallback(
    async (notificationId: string) => {
      if (!userId) return;

      try {
        // Optimistically update the UI
        setNotifications((prev) =>
          prev.map((n) =>
            n.id === notificationId ? { ...n, is_read: true } : n,
          ),
        );
        setUnreadCount((prev) => Math.max(0, prev - 1));
      } catch (error) {
        console.error("Error in markAsRead:", error);
      }
    },
    [userId],
  );

  const addNotification = useCallback((notification: Notification) => {
    setNotifications((prev) => [notification, ...prev]);
    setUnreadCount((prev) => prev + 1);
  }, []);

  return {
    notifications,
    unreadCount,
    markAsRead,
    addNotification, // Expose addNotification to add new notifications
  };
}
