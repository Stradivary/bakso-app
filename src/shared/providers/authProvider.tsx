import { Session } from '@supabase/supabase-js';
import { PropsWithChildren, useCallback, useEffect, useState } from 'react';
import { Database } from '../models/supabase';
import { signInUser } from '../services/authService';
import { supabase } from '../services/supabaseService';
import { AuthContext } from '../hooks/useAuth';

type UserDetails = Database['public']['Tables']['user_profiles']['Row'];

export interface AuthContextType {
  session: Session | null;
  userDetails: UserDetails | null;
  isLoading: boolean;
  error: string | null;
  login: (name: string, role: string, latitude: number, longitude: number) => Promise<void>;
  logout: () => Promise<void>;
  isAuthenticated: boolean;
}

export const AuthProvider: React.FC<PropsWithChildren> = ({ children }) => {
  const [session, setSession] = useState<Session | null>(null);
  const [isLoading, setIsLoading] = useState(true);
  const [error, setError] = useState<string | null>(null);

  const login = async (name: string, role: string, latitude: number, longitude: number) => {
    setIsLoading(true);
    try {
      const { session, error } = await signInUser(name, role, latitude, longitude);
      if (error) throw error;
      setSession(session);
      setError(null);
    } catch (error) {
      console.error('Error during login:', error);
      setError('Failed to login');
    } finally {
      setIsLoading(false);
    }
  };

  const logout = async () => {
    try {
      await supabase.auth.signOut();
      setSession(null);
      setError(null);
    } catch (error) {
      console.error('Error during logout:', error);
      setError('Failed to logout');
    }
  };

  const handleAuthStateChange = useCallback((_event: string, newSession: Session | null) => {
    setSession(newSession);
  }, []);

  useEffect(() => {
    const initializeAuth = async () => {
      setIsLoading(true);
      try {
        const { data: { session: initialSession }, error } = await supabase.auth.getSession();
        if (error) {
          console.error('Error getting initial session:', error);
          setError('Failed to get initial session');
        } else {
          setSession(initialSession);
        }
      } catch (error) {
        console.error('Error getting initial session:', error);
        setError('Failed to get initial session');
      } finally {
        setIsLoading(false);
      }
    };

    initializeAuth();

    const { data: { subscription } } = supabase.auth.onAuthStateChange(handleAuthStateChange);

    return () => {
      subscription.unsubscribe();
    };
  }, [handleAuthStateChange]);

  const isAuthenticated = !!session;

  return (
    <AuthContext.Provider value={{
      session,
      userDetails: null,
      isLoading,
      error,
      login,
      logout,
      isAuthenticated
    }}>
      {children}
    </AuthContext.Provider>
  );
};
