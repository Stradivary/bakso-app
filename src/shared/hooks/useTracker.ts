import {
  calculateRegion,
  filterNearbyUsers,
  initSupabaseChannel,
  User,
} from "@/shared/services/trackerServices";
import { RealtimeChannel } from "@supabase/supabase-js";
import { LatLng } from "leaflet";
import { useCallback, useEffect, useRef, useState } from "react";
import { createPingPayload } from "../services/trackerServices";
import { Notification, useNotifications } from "./useNotification";

export const useTracker = (
  userId: string,
  userRole: "seller" | "buyer",
  initialLocation: LatLng,
  userName: string,
) => {
  const [nearbyUsers, setNearbyUsers] = useState<User[]>([]);
  const [selectedSeller, setSelectedSeller] = useState<string | null>(null);
  const [localNotifications, setLocalNotifications] = useState<Notification[]>(
    [],
  );
  const { markAsRead } = useNotifications(userId, localNotifications);

  const regionRef = useRef<string>("");
  const channelRef = useRef<RealtimeChannel | null>(null);

  const handleUnsubscribe = useCallback(() => {
    channelRef.current?.untrack();
    channelRef.current?.unsubscribe();
  }, []);

  const handlePresenceSync = useCallback(
    (users: User[]) => {
      const filteredUsers = filterNearbyUsers(
        users,
        initialLocation,
        userRole,
        userId,
      );
      setNearbyUsers(filteredUsers);
    },
    [initialLocation, userRole, userId],
  );

  const handlePing = useCallback(
    (buyerId: string, buyerName: string) => {
      const notification = createPingPayload({
        buyer_id: buyerId,
        buyer_name: buyerName,
        user_id: userId,
      });
      setLocalNotifications((prev) => [...prev, notification]);
      setSelectedSeller(userId);
    },
    [userId],
  );

  const handleLocationUpdate = useCallback(
    (userId: string, location: LatLng) => {
      setNearbyUsers((prev) =>
        prev.map((user) => (user.id === userId ? { ...user, location } : user)),
      );
    },
    [],
  );

  useEffect(() => {
    const region = calculateRegion(initialLocation.lat, initialLocation.lng);
    regionRef.current = region;

    initSupabaseChannel(
      initialLocation,
      region,
      {
        userId,
        userRole,
        userName,
      },
      {
        handleLocationUpdate,
        handlePing,
        handlePresenceSync,
      },
    ).then((channel) => {
      channelRef.current = channel;
    });

    return () => {
      if (channelRef.current) {
        channelRef.current.unsubscribe();
      }
    };
  }, [
    userId,
    userRole,
    userName,
    initialLocation,
    handlePresenceSync,
    handlePing,
    handleLocationUpdate,
  ]);

  return {
    nearbyUsers,
    selectedSeller,
    markAsRead,
    handlePing,
    handleLocationUpdate,
    notifications: localNotifications,
    deactivateUser: handleUnsubscribe,
  };
};
