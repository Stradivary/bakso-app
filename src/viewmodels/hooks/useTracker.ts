import { RealtimeChannel } from "@supabase/supabase-js";
import { LatLng } from "leaflet";
import { useCallback, useEffect, useRef, useState } from "react";
import { useNotifications } from "./useNotification";
import { filterNearbyUsers } from "@/shared/utils/filterNearbyUsers";
import { createPingPayload } from "@/shared/utils/createPingPayload";
import { calculateRegion } from "@/shared/utils/calculateRegion";
import { initSupabaseChannel } from "@/viewmodels/services/trackerServices";
import { UserPayload } from "@/models/user.types";

export const useTracker = (
  userId: string,
  userRole: "seller" | "buyer",
  initialLocation: LatLng | null,
  userName: string,
) => {
  const [nearbyUsers, setNearbyUsers] = useState<UserPayload[]>([]);
  const { notifications, markAsRead, addNotification } =
    useNotifications(userId);

  const regionRef = useRef<string>("");
  const channelRef = useRef<RealtimeChannel | null>(null);

  const handleUnsubscribe = useCallback(() => {
    channelRef.current?.untrack();
    channelRef.current?.unsubscribe();
  }, []);

  const handlePresenceSync = useCallback(
    (users: UserPayload[]) => {
      if (!initialLocation) return users;
      const filteredUsers = filterNearbyUsers(
        users,
        initialLocation,
        userRole,
        userId,
      );
      setNearbyUsers(filteredUsers);
    },
    [userRole, userId, initialLocation],
  );

  // Function to handle receiving pings (for sellers)
  const handlePing = useCallback(
    (buyerId: string, buyerName: string) => {
      const notification = createPingPayload({
        buyer_id: buyerId,
        buyer_name: buyerName,
        user_id: userId,
      });
      addNotification(notification);
    },
    [userId, addNotification],
  );

  // Function to send pings (for buyers)
  const sendPing = useCallback(
    (sellerId: string) => {
      if (channelRef.current) {
        channelRef.current.send({
          type: "broadcast",
          event: "ping",
          payload: {
            seller_id: sellerId,
            buyer_id: userId,
            buyer_name: userName,
          },
        });
      }
    },
    [userId, userName],
  );

  const handleLocationUpdate = useCallback(
    (userId: string, location: LatLng) => {
      setNearbyUsers((prev) =>
        prev.map((user) => (user.id === userId ? { ...user, location } : user)),
      );
    },
    [],
  );

  useEffect(() => {
    if (!initialLocation) return;
    const region = calculateRegion(initialLocation.lat, initialLocation.lng);
    regionRef.current = region;
    initSupabaseChannel(
      initialLocation,
      region,
      {
        userId,
        userRole,
        userName,
      },
      {
        handleLocationUpdate,
        handlePing,
        handlePresenceSync,
      },
    ).then((channel) => {
      channelRef.current = channel;
    });

    return () => {
      if (channelRef.current) {
        channelRef.current.unsubscribe();
      }
    };
  }, [
    userId,
    userRole,
    userName,
    initialLocation,
    handlePresenceSync,
    handlePing,
    handleLocationUpdate,
  ]);

  return {
    nearbyUsers,
    sendPing,
    handleLocationUpdate,
    notifications,
    markAsRead,
    deactivateUser: handleUnsubscribe,
  };
};
