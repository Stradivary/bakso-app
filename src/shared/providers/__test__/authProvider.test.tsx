import { render, screen, } from '@testing-library/react';
import { describe, it, beforeEach, vi, expect } from 'vitest';
import { AuthProvider } from '../authProvider';
import { AuthContext } from '@/shared/hooks/useAuth';

vi.mock('../services/authService');

describe('AuthProvider', () => {
    beforeEach(() => {
        vi.clearAllMocks();
    });

    it('should initialize with loading state', async () => {
        render(
            <AuthProvider>
                <div>Test</div>
            </AuthProvider>
        );

        expect(screen.getByText('Test')).toBeDefined();
    });
 
    it('should handle session integrity validation', async () => {
        render(
            <AuthProvider>
                <div>Test</div>
            </AuthProvider>
        );

    });
});