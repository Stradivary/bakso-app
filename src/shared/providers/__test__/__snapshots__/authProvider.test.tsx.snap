// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AuthProvider > renders correctly and matches snapshot 1`] = `
<div>
  <div>
    Test Child
  </div>
</div>
`;
